tests/integration/web/management_profile_test.py

```diff
@@ -0,0 +1,119 @@
+import pytest
+import pytest_twisted
+from django.urls import reverse
+
+from nav.ipdevpoll.plugins.paloaltoarp import PaloaltoArp
+from nav.models.manage import ManagementProfile, Netbox
+from nav.web.seeddb.page.management_profile.forms import HttpApiForm
+
+
+class TestHttpApiForm:
+    class TestServiceChoices:
+        """
+        The HttpApiForm Django form, which creates HTTP API management profiles,
+        stores a "service" key in the profile's configuration, used by other
+        parts of NAV to differentiate between HTTP services:
+
+        profile.configuration == {"service": "foo", ...}
+
+        The HttpApiForm follows the convention of other management profile forms
+        in not using named constants for ChoiceField values stored in the
+        configuration dict. To help make sure that other parts of NAV use the
+        same service values as those set by the form, we add tests below for
+        each service choice.
+
+        The tests first associates a service with a netbox through using the
+        form, thus asserting that the service is a valid choice. Then it checks
+        that the other parts of NAV accepts same service value as the one the
+        form accepted.
+        """
+
+        tested_services = ["Palo Alto ARP"]
+
+        @pytest.mark.twisted
+        @pytest_twisted.inlineCallbacks
+        def test_paloalto_plugin_should_use_the_service_choice_used_in_form(
+            self, localhost, send_add_management_profile_form
+        ):
+            can_handle = yield PaloaltoArp.can_handle(localhost)
+            assert not can_handle
+
+            # Check that "Palo Alto ARP" is accepted by the form
+            send_add_management_profile_form(
+                localhost,
+                protocol=ManagementProfile.PROTOCOL_HTTP_API,
+                service="Palo Alto ARP",
+                api_key="foo",
+            )
+
+            # Check that "Palo Alto ARP" is accepted by the plugin
+            can_handle = yield PaloaltoArp.can_handle(localhost)
+            assert can_handle
```

Comment:
Is it *always* a bad idea to have some integration tests tend towards more elaborate end-to-end tests😅? In the original pull-request I had a test similar to the (new) one pasted below, which was suggested to be split up into smaller tests with e.g. one part placed in `paloaltoarp_test.py` and the other placed in `management_profile_test.py`. Splitting the tests makes it easier to parse test results and pinpoint errors, but no matter what I do, I find the tests seem a bit out of place that way, and I find it becomes harder to reason about the intent and necessity of the tests when they are separated. I've removed all the `management_profile_test.py` tests from previous commits and replaced the tests with a single end-to-end test in `paloaltoarp_test.py` instead.

This test catches erroneous assumptions the plugin may make about what data the form accepts and puts into the respective profile's configuration dictionary. This could be done in a less "end-to-end way" by making assertions on the concrete values supplied to and created by the form (as opposed to just checking that the plugin and form behaves well with these values), but such a test would need to be updated along with the plugin's implementation to a higher degree, and reading the test's code alone wouldn't give much of a clue what parts of the plugin is tested.

https://github.com/Uninett/nav/blob/c930c2f29efd399cf1a227a8246281d45b1534ba/tests/integration/ipdevpoll/plugins/paloaltoarp_test.py#L111-L179

Reply:
> Is it _always_ a bad idea to have some integration tests tend towards more elaborate end-to-end tests😅?

Not at all - but you should have smaller unit tests for the individual parts as well.

The larger the test, the harder it is to reason about what is actually broken when it fails.  This is also why it is ideal to strive to have only 1 single assert in a test: If you have multiple asserts in a test, it is often a sign that you should break it into multiple tests.

If your unit tests are structured properly, you might find that a full-on end-to-end test isn't strictly necessary, because it will not cover any issues that the unit tests will not already cover.  But there is no hard-and-fast rule for these kinds of things, and I don't think having test redundancies is bad (except you now have more test code to maintain as well).

Reply:
> but such a test would need to be updated along with the plugin's implementation,

Well, if you change the behavior, you also need to change the tests.  If you are just refactoring but keeping the behavior, tests shouldn't need to change (really, if the test needs to change due to refactoring, you're not really refactoring in the strict sense of the word - or your test is b0rked to begin with)